### ISSUE TITLE
Some timezones breaks after migrating from pytz to zoneinfo

### ISSUE THREAD

**Issue by tobixen** [collective/icalendar#333]
I just encountered a strange issue when running tests in my [calendar-cli](https://github.com/tobixen/calendar-cli/) project. In particular, it broke with the DeNoronha timestamp (according to a contributor of test code to the [caldav library](https://github.com/python-caldav/caldav/), this place is well-known for being a "remote place", so it's used as a "remote time zone" in the tests).  After migrating from pytz to zoneinfo the test code started breaking.

I wrote up the following code to highlight the breakage:

```python
import datetime
import pytz
import zoneinfo
import icalendar

utc1 = pytz.utc
utc2 = zoneinfo.ZoneInfo('UTC')
oslo1 = pytz.timezone('Europe/Oslo')
oslo2 = zoneinfo.ZoneInfo('Europe/Oslo')
remote1 = pytz.timezone('Brazil/DeNoronha')
remote2 = zoneinfo.ZoneInfo('Brazil/DeNoronha')

for tz in (utc1, utc2, oslo1, oslo2, remote1, remote2):
    myevent = icalendar.Event()
    myevent.add('dtstart', datetime.datetime.now().astimezone(tz))
    print(myevent.to_ical().decode('ASCII'))
```

When running it, I get the following output:

```
BEGIN:VEVENT
DTSTART;VALUE=DATE-TIME:20211116T232629Z
END:VEVENT

BEGIN:VEVENT
DTSTART;TZID=UTC;VALUE=DATE-TIME:20211116T232629Z
END:VEVENT

BEGIN:VEVENT
DTSTART;TZID=Europe/Oslo;VALUE=DATE-TIME:20211117T002629
END:VEVENT

BEGIN:VEVENT
DTSTART;TZID=CET;VALUE=DATE-TIME:20211117T002629
END:VEVENT

BEGIN:VEVENT
DTSTART;TZID=Brazil/DeNoronha;VALUE=DATE-TIME:20211116T212629
END:VEVENT

BEGIN:VEVENT
DTSTART;TZID=-02;VALUE=DATE-TIME:20211116T212629
END:VEVENT
```

Note that for every time zone above, the output has changed simply from moving from pytz to zoneinfo.  I would expect there to be no changes in the output, and I think the pytz behaviour is correct:
* for timestamp suffixed with Zulu,  no point to specify UTC.
* The full name of the timezone should be given, and not the short name
* In particular, "-02" is simply not an acceptable TZID.

**Comment by tobixen** [comment]
I entered debug mode and found a function `tzid_from_dt` that returns the wrong thing

```
(Pdb) tzid_from_dt(pytz.timezone('Europe/Oslo').localize(datetime.now()))
'Europe/Oslo'
(Pdb) tzid_from_dt(datetime.now().astimezone(zoneinfo.ZoneInfo('Europe/Oslo')))
'CET'
```

Unfortunately, I have to run now, won't get time to look more into this.

**Comment by tobixen** [comment]
I found the `tzid_from_dt` and managed to mend it so the new test code passes.

However, it will still be broken for time zones created through `dateutil.tz`.  There seems to be no trivial way to find the full name of the time zone from a given timezone object produced by `dateutil.tz`.

Timestamps and time zones in Python is such a mess ...

**Comment by tobixen** [comment]
New test run of the code above:

```
$ python ./test-ical.py
BEGIN:VEVENT
DTSTART;VALUE=DATE-TIME:20211117T150915Z
END:VEVENT

BEGIN:VEVENT
DTSTART;TZID=UTC;VALUE=DATE-TIME:20211117T150915Z
END:VEVENT

BEGIN:VEVENT
DTSTART;TZID=Europe/Oslo;VALUE=DATE-TIME:20211117T160915
END:VEVENT

BEGIN:VEVENT
DTSTART;TZID=Europe/Oslo;VALUE=DATE-TIME:20211117T160915
END:VEVENT

BEGIN:VEVENT
DTSTART;TZID=Brazil/DeNoronha;VALUE=DATE-TIME:20211117T130915
END:VEVENT

BEGIN:VEVENT
DTSTART;TZID=Brazil/DeNoronha;VALUE=DATE-TIME:20211117T130915
END:VEVENT
```

Still one issue here, the UTC time zone should not be displayed.  I'll create a separate issue for that.

Replacing zoneinfo.ZoneInfo with dateutil.tz.gettz, the original issues still remain.  I'll create a separate issue for that, too.

**Comment by NicoHood** [comment]
Does this library use pytz or zoneinfo? Because in the readme it states pytz and I was about to ask, if we can replace it with zoneinfo nowadays?

**Comment by tobixen** [comment]
Zoneinfo is the way ahead, pytz is legacy.

**Comment by tobixen** [comment]
I believe this bug can be closed, as #339 has been merged.  Or should the issue stay open until a proper release has been done?

**Comment by niccokunzmann** [comment]
Hi, thanks again! I will close this when issues get merged because this is for development and the changelog for the releases. Thanks for asking!
